Address Eric Crockett's review comments.

chris-wood opened the pull request:
cc @crockeea

chris-wood commented:
@hugokraw this PR should be good to go!

hugokraw commented:
I am afraid that I did not pushed  my comments (which I wrote them a couple of days ago) before for you to see.  Sorry for that. Eric also asked for a couple of clarifications to add (on the need of forward secrecy and something else). If you need text from me let me know.

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -350,8 +354,9 @@ OPAQUE relies on the following protocols and primitives:
 
-We also assume the existence of a function `KeyGen`, which generates an OPRF private
-and public key. We write `(skU, pkU) = KeyGen()` to denote this function.
+We also assume the existence of a function `KeyGen` from {{I-D.irtf-cfrg-voprf}}, which
```
It's currently not part of the OPRF API, so I left it off the list above. If the OPRF draft changes, we can change this too.

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -362,19 +367,13 @@ them as input to an authenticated key exchange (AKE) protocol.
-outside the scope of this document, though may be addressed elsewhere, such as
-in {{I-D.sullivan-tls-opaque}}.
+We first define the core OPAQUE protocol based on a generic OPRF, hash, and MHF function.
+Specification of the AKE is out of scope for this document. {{instantiations}} describes
```
What I meant is that this document doesn't intend to specify, describe, etc *new* AKEs, but I see how this is confusing. I'll just remove it.

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -876,8 +875,8 @@ Steps:
 
 As in the registration phase, applications MUST authenticate pkS; secrecy of pkS is
-optional. If an application requires secrecy of pkS, this value SHOULD be omitted
-from auth_data (step 9).
```
I think this entire paragraph is probably obsolete. I'll just remove it.

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -898,37 +897,37 @@ OPRF computation and credential fetch flow.
-denoted KE1, KE2, KE3, where KE1 and KE2 include DH values sent by client and
-server, respectively, and KE3 provides explicit client authentication and full
+This section describes several instantiations of OPAQUE using different AKE protocols.
+As discussed in {{security-considerations}}, each AKE protocol requires KCI resistance.
```
That's a nice rewrite -- I'll take it!

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -688,8 +686,9 @@ is that which is associated with the OPAQUE configuration (see {{configurations}
-pkU, IdU, or IdS in the `Credentials.secret_credentials` structure (step 5) if secrecy
+`Credentials.secret_credentials` list (step 5), and MUST NOT be included in the
+`Credentials.cleartext_credentials` list. Applications may optionally include
+pkU, idU, or IdS in the `Credentials.secret_credentials` structure (step 5) if secrecy
```
Fixed!

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -748,10 +747,10 @@ the OPRF protocol.
 
 Note also that the authenticate stage can run the OPRF and AKE protocols concurrently
 with interleaved and combined messages (while preserving the internal ordering of
```
Fixed!

chris-wood commented:
@hugokraw Thanks for the feedback -- I incorporated your comments. I also added some additional changes on top. Namely:

- Fix the auth_tag size to the output of HMAC. (One less variable-length thing to worry about.)
- Change how the auth_tag is computed. Previously, it did not include the length of the ciphertext or auth_data, which was worrisome. Now it's computed over the structure:

```
struct {
  opaque nonce[32];
  opaque ct<1..2^16-1>;
  opaque auth_data<0..2^16-1>;
} InnerEnvelope;
```

which includes the length of `ct` and `auth_data`.

Please let me know what you think!

chris-wood commented:
@hugokraw, friendly bump! Please feel free to edit the PR directly if that's easier for you. :-)

hugokraw commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -1406,60 +1400,60 @@ whether a given user identity is registered with a server. Preventing
+server can run an exhaustive offline dictionary attack to validate guesses for the user's
+password; this is inevitable in any aPAKE protocol. (OPAQUE enables a defense against such
+offline dictionary attacks by distributing the server so that an offline attack is only
+possible if all - or a minimal number of - servers are compromised.)
```
Add a reference to the OPAQUE paper at the end of the parentheses

hugokraw commented:
Sorry for the delay in responding to the latest changes. I was busy welcoming my second grandson... :-)
